server/handler/push.go

```diff
@@ -80,6 +80,10 @@ func (h *Push) Handle(ctx context.Context, eventType, deliveryID string, payload
 		if err := h.UpdatePullRequest(logger.WithContext(ctx), pullCtx, client, pr, baseRef); err != nil {
```

Comment:
I think it would be a good idea to change `UpdatePullRequest` to return `(bool, error)`, with the first value indicating if an update was started or not. That way, only PRs that are _not_ updated will be processed for merging; the update process should trigger the normal hooks in the other case.

Reply:
`UpdatePullRequest` only starts the update, it does not check that the branch is actually behind the base, so the following happens:

Event 1 (PR branch is behind base):
- `UpdatePullRequest` returns `true`, as PR can be updated
- Merge is not performed
- Update is performed asynchronously.

Event 2 (PR branch is up-to-date):
- `UpdatePullRequest` still returns `true`, as branch still can be updated
- Merge is not performed
- Update is skipped asynchronously, as branch is up-to-date.

The only way to sequence these operations is to wait for update to check whether it's behind the base or not. May not be a bad idea, but needs some more refactoring.

Reply:
That's fair, I forgot that `UpdatePullRequest` only checks if updates are enabled, not if an update will actually be made.